### PR TITLE
add firestore support for heartbeat and user agent string

### PR DESCRIPTION
**Pull request by VinayGuthal**
This pr adds in an interface grpcMetadata which the FirestoreChannel uses to populate its headers. This also adds in the UserAgentInformation and the HeartBeatInformation for usage in the firestore backend.

### CODE REVIEW

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirestoreChannel.java [hunk 1]
```diff
@@ -283,6 +288,7 @@ private Metadata requestHeaders() {
     headers.put(X_GOOG_API_CLIENT_HEADER, X_GOOG_API_CLIENT_VALUE);
     // This header is used to improve routing and project isolation by the backend.
     headers.put(RESOURCE_PREFIX_HEADER, this.resourcePrefixValue);
+    grpcMetadata.updateMetadata(headers);
```
Comment (VinayGuthal) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirestoreChannel.java]: add nullability check

File: firebase-firestore/src/main/java/com/google/firebase/firestore/grpc/DefaultGrpcMetadata.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import com.google.firebase.heartbeatinfo.HeartBeatInfo;
+import com.google.firebase.platforminfo.UserAgentPublisher;
+import io.grpc.Metadata;
+
+/**
+ * Class provides an implementation of the GrpcMetadata interface.
+ *
+ * <p>This updates the metadata with platformInfo string and the heartBeatInfo code.
+ */
+public class DefaultGrpcMetadata implements GrpcMetadata {
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/grpc/DefaultGrpcMetadata.java]: `DefaultGrpcMetadata` is not very descriptive. Since this is related to `x-firebase-client*` headers could we use `FirebaseClient` instead of `Default`?

Other alternatives:
  * `HeartBeat`-  -- though this isn't quite right because this handles both the user agent and heartbeat headers.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java [hunk 3]
```diff
@@ -41,13 +44,17 @@
         Component.builder(FirestoreMultiDbComponent.class)
             .add(Dependency.required(FirebaseApp.class))
             .add(Dependency.required(Context.class))
+            .add(Dependency.required(HeartBeatInfo.class))
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java]: These should be optional dependencies, shouldn't they?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/grpc/DefaultGrpcMetadata.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  public DefaultGrpcMetadata(
+      @NonNull UserAgentPublisher userAgentPublisher, @NonNull HeartBeatInfo heartBeatInfo) {
+    this.userAgentPublisher = userAgentPublisher;
+    this.heartBeatInfo = heartBeatInfo;
+  }
+
+  @Override
+  public void updateMetadata(@NonNull Metadata metadata) {
+    metadata.put(USER_AGENT_HEADER, userAgentPublisher.getUserAgent());
```
[the remark this answers is not in the document]
Comment (VinayGuthal) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/grpc/DefaultGrpcMetadata.java]: The value of these headers would dictate the type of heartbeat.
0 - no heartbeat
1 - sdk heartbeat
2- global heartbeat
3 - combined heartbeat.
I can put the condition to put the header only if the value is not 0. Is there any problem in sending the value anyway? 
The backend code will handle the scenarios when the value is 1, 2 or 3 and assume 0 means no heart beat.
. The USER_AGENT_HEADER needs to be added unconditionally

File: firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java (hunk 3 above)
Comment (VinayGuthal) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java]: These should be required? Bcz this assumes we have the firebase dependencies during component initialization. The constructor should have nulllable attributes so that they can be used in the tests. Am I understanding this currently?

cc @vkryachko

File: firebase-firestore/src/main/java/com/google/firebase/firestore/grpc/DefaultGrpcMetadata.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/grpc/DefaultGrpcMetadata.java]: Let's discuss offline.

[the remark this answers is not in the document]
Comment (VinayGuthal) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/grpc/DefaultGrpcMetadata.java]: Have updated the heartbeatinfo header to be sent only once a day. Once I discuss with you we can decide as to what to do about userplatforminfo

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java [hunk 5]
```diff
@@ -45,8 +45,12 @@ public DefaultGrpcMetadata(
   @Override
   public void updateMetadata(@NonNull Metadata metadata) {
     metadata.put(USER_AGENT_HEADER, userAgentPublisher.getUserAgent());
-    metadata.put(
-        HEART_BEAT_HEADER,
-        Integer.toString(heartBeatInfo.getHeartBeatCode(firebaseFirestoreHeartBeatTag).getCode()));
+    int heartBeatCode = heartBeatInfo.getHeartBeatCode(firebaseFirestoreHeartBeatTag).getCode();
+    if (heartBeatCode != 0) {
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: Since we can't link to the design doc for this here, could you add a comment about why `0` is special here?

```
// Non-zero values indicate some kind of heartbeat should be sent
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java [hunk 6]
```diff
@@ -45,8 +45,12 @@ public DefaultGrpcMetadata(
...
+      metadata.put(
+          HEART_BEAT_HEADER,
+          Integer.toString(
+              heartBeatInfo.getHeartBeatCode(firebaseFirestoreHeartBeatTag).getCode()));
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: Re-use `heartBeatCode` here? Otherwise it's very likely that you'll end up sending zero every time here since the first call consumed the beat for the day.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java [hunk 7]
```diff
@@ -45,8 +45,12 @@ public DefaultGrpcMetadata(
   @Override
   public void updateMetadata(@NonNull Metadata metadata) {
     metadata.put(USER_AGENT_HEADER, userAgentPublisher.getUserAgent());
-    metadata.put(
-        HEART_BEAT_HEADER,
-        Integer.toString(heartBeatInfo.getHeartBeatCode(firebaseFirestoreHeartBeatTag).getCode()));
+    int heartBeatCode = heartBeatInfo.getHeartBeatCode(firebaseFirestoreHeartBeatTag).getCode();
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: How costly is calling `getHeartBeatCode`? Does it perform any I/O? Does it talk to the network?

Comment (VinayGuthal) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: it looks into the file stored in the app to check when the last time the heartbeat was sent. It doesnt talk to network.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java (hunk 6 above)
Comment (VinayGuthal) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: reused nice catch!

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java (hunk 5 above)
Comment (VinayGuthal) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: yes sure! Thanks!

File: firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java (hunk 3 above)
Comment (vkryachko) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java]: Logically it seems optional, as it is not critical to the operation of Firestore, however I don't think it matters much as the components in question are part of `firebase-common` so there should not be a case where they would be missing at runtime, this could in theory change in the future though.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java [hunk 8]
```diff
@@ -41,13 +44,17 @@
...
+            .add(Dependency.required(UserAgentPublisher.class))
             .add(Dependency.optional(InternalAuthProvider.class))
             .factory(
                 c ->
                     new FirestoreMultiDbComponent(
                         c.get(Context.class),
                         c.get(FirebaseApp.class),
-                        c.get(InternalAuthProvider.class)))
+                        c.get(InternalAuthProvider.class),
+                        new FirebaseClientGrpcMetadataProvider(
```
Comment (vkryachko) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java]: If you choose to make the components optional pls update this code to check for nulls and inject null instead of `FirebaseClientGrpcMetadataProvider`

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java (hunk 7 above)
Comment (vkryachko) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: it uses `SharedPreferences` so technically it only reads from disk once and keeps/manipulates state in memory, syncing to disk asynchronously.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java (hunk 3 above)
Comment (vkryachko) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java]: Pls switch to provider dependency, i.e. `requiredProvider()`/`optionalProvider()`, the components are not needed until later, so it's nice to not do more work than necessary(initializing the components) during Firestore initialization.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java [hunk 9]
```diff
@@ -41,13 +44,17 @@
...
+                            c.get(UserAgentPublisher.class), c.get(HeartBeatInfo.class))))
```
Comment (vkryachko) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/FirestoreRegistrar.java]: This needs to change to `c.getProvider()`, and also `FirebaseClientGrpcMetadataProvider` needs to change to use `Provider<T>` instead of values directly

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProviderTest.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    metadataProvider.updateMetadata(metadata);
+    assertThat(metadata.keys().size()).isEqualTo(0);
+  }
+
+  @Test
+  public void updateHeaderWhenHBCodeisNoneZero() {
+    Metadata metadata = new Metadata();
+    when(mockUserAgentProvider.get()).thenReturn(mockUserAgent);
+    when(mockHeartBeatProvider.get()).thenReturn(mockHeartBeat);
+    when(mockHeartBeat.getHeartBeatCode(any())).thenReturn(HeartBeatInfo.HeartBeat.GLOBAL);
```
Comment (vkryachko) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProviderTest.java]: pls add test for `HeartBeat.SDK`

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProviderTest.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      Metadata.Key.of("x-firebase-client-log-type", Metadata.ASCII_STRING_MARSHALLER);
+
+  private static final Metadata.Key<String> USER_AGENT_HEADER =
+      Metadata.Key.of("x-firebase-client", Metadata.ASCII_STRING_MARSHALLER);
+
+  @Test
+  public void noUpdateWhenNullProvider() {
+    Metadata metadata = new Metadata();
+    when(mockUserAgentProvider.get()).thenReturn(null);
+    when(mockHeartBeatProvider.get()).thenReturn(null);
```
Comment (vkryachko) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProviderTest.java]: pls add 2 tests when only one of the providers is null

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+package com.google.firebase.firestore.remote;
+
+import androidx.annotation.NonNull;
+import com.google.firebase.heartbeatinfo.HeartBeatInfo;
+import com.google.firebase.inject.Provider;
+import com.google.firebase.platforminfo.UserAgentPublisher;
+import io.grpc.Metadata;
+
+/**
+ * Class provides an implementation of the GrpcMetadataProvider interface.
```
Comment (vkryachko) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]:
```suggestion
 * Provides an implementation of the GrpcMetadataProvider interface.
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  public FirebaseClientGrpcMetadataProvider(
+      @NonNull Provider<UserAgentPublisher> userAgentPublisherProvider,
+      @NonNull Provider<HeartBeatInfo> heartBeatInfoProvider) {
+    this.userAgentPublisherProvider = userAgentPublisherProvider;
+    this.heartBeatInfoProvider = heartBeatInfoProvider;
+  }
+
+  @Override
+  public void updateMetadata(@NonNull Metadata metadata) {
+    if (heartBeatInfoProvider.get() == null || userAgentPublisherProvider.get() == null) {
```
Comment (vkryachko) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: would it make sense to still send heartbeat if userAgentPublisher is null?

File: tools/measurement/apksize/src/inappmessagingdisplay/java/com.google.apksize/InAppMessagingDisplay.java [hunk 14]
```diff
@@ -16,16 +16,16 @@
 
 import android.app.Activity;
 import android.content.Context;
+import androidx.annotation.NonNull;
```
Comment (vkryachko) [inline tools/measurement/apksize/src/inappmessagingdisplay/java/com.google.apksize/InAppMessagingDisplay.java]: Is this file change intentional?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java (hunk 13 above)
Comment (VinayGuthal) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirebaseClientGrpcMetadataProvider.java]: no i dont think so... Bcz with the userAgentPublisher is how we know which sdk is being used.

File: tools/measurement/apksize/src/inappmessagingdisplay/java/com.google.apksize/InAppMessagingDisplay.java (hunk 14 above)
Comment (VinayGuthal) [reply tools/measurement/apksize/src/inappmessagingdisplay/java/com.google.apksize/InAppMessagingDisplay.java]: this is a part of googleJavaFormat

Comment (VinayGuthal) [reply tools/measurement/apksize/src/inappmessagingdisplay/java/com.google.apksize/InAppMessagingDisplay.java]: it just organized the imports

File: firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java [hunk 15]
```diff
@@ -85,10 +87,12 @@ public FirestoreClient(
       DatabaseInfo databaseInfo,
       FirebaseFirestoreSettings settings,
       CredentialsProvider credentialsProvider,
-      final AsyncQueue asyncQueue) {
+      final AsyncQueue asyncQueue,
+      @Nullable GrpcMetadataProvider metadata) {
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/core/FirestoreClient.java]: `metadataProvider`

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/Datastore.java [hunk 16]
```diff
@@ -87,12 +88,15 @@ public Datastore(
       DatabaseInfo databaseInfo,
       AsyncQueue workerQueue,
       CredentialsProvider credentialsProvider,
-      Context context) {
+      Context context,
+      @Nullable GrpcMetadataProvider grpcMetadataProvider) {
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/Datastore.java]: This could be `metadataProvider` like elsewhere.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirestoreChannel.java [hunk 17]
```diff
@@ -64,12 +64,16 @@
...
 
+  private final GrpcMetadataProvider grpcMetadataProvider;
+
   FirestoreChannel(
       AsyncQueue asyncQueue,
       Context context,
       CredentialsProvider credentialsProvider,
-      DatabaseInfo databaseInfo) {
+      DatabaseInfo databaseInfo,
+      GrpcMetadataProvider metadata) {
```
Comment (wilhuff) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/FirestoreChannel.java]: `metadataProvider` and possibly make `this.grpcMetadataProvider` match.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestore.java [hunk 18]
```diff
@@ -109,7 +111,8 @@ static FirebaseFirestore newInstance(
       @NonNull FirebaseApp app,
       @Nullable InternalAuthProvider authProvider,
       @NonNull String database,
-      @NonNull InstanceRegistry instanceRegistry) {
+      @NonNull InstanceRegistry instanceRegistry,
+      @Nullable GrpcMetadataProvider metadata) {
```
Comment (vkryachko) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestore.java]:
```suggestion
      @Nullable GrpcMetadataProvider metadataProvider) {
```

File: firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestore.java [hunk 19]
```diff
@@ -134,7 +137,7 @@ static FirebaseFirestore newInstance(
 
     FirebaseFirestore firestore =
         new FirebaseFirestore(
-            context, databaseId, persistenceKey, provider, queue, app, instanceRegistry);
+            context, databaseId, persistenceKey, provider, queue, app, instanceRegistry, metadata);
```
Comment (vkryachko) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/FirebaseFirestore.java]:
```suggestion
            context, databaseId, persistenceKey, provider, queue, app, instanceRegistry, metadataProvider);
```
